modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/consensus/ConsensusStateUpdater.scala

```diff
@@ -265,35 +267,43 @@ object ConsensusStateUpdater {
                     }.filter { case (_, events) => events.nonEmpty }
                     _ <- consensusStorage.addEvents(returnedPeerEvents)
                     hash <- artifact.hashF
-                    effect = gossip.spread(ConsensusPeerDeclaration(key, Proposal(hash)))
-                  } yield (ProposalInfo[Artifact](artifact, hash).some, effect)
-                }.map(_.getOrElse((none[ProposalInfo[Artifact]], Applicative[F].unit))).map {
-                  case (maybeProposalInfo, effect) =>
-                    val newState = state.copy(status = CollectingProposals[Artifact](majorityTrigger, maybeProposalInfo))
+                    effect = gossip.spread(ConsensusPeerDeclaration(key, Proposal(hash))) *>
+                      gossip.spreadCommon(ConsensusArtifact(key, artifact))
+                  } yield (ProposalInfo[Artifact](artifact, hash).some, facilityInfo.lastSignedArtifact.some, effect)
+                }.map(_.getOrElse((none[ProposalInfo[Artifact]], none[Signed[Artifact]], Applicative[F].unit))).map {
+                  case (maybeProposalInfo, maybeLastSignedArtifact, effect) =>
+                    val newState =
+                      state.copy(status = CollectingProposals[Artifact](majorityTrigger, maybeProposalInfo, maybeLastSignedArtifact))
                     (newState, effect)
                 }
           }
-        case CollectingProposals(majorityTrigger, maybeProposalInfo) =>
+
+        case CollectingProposals(majorityTrigger, maybeProposalInfo, maybeLastArtifact) =>
           val maybeAllProposals = state.facilitators
             .traverse(resources.peerDeclarationsMap.get)
             .flatMap(_.traverse(_.proposal.map(_.hash)))
 
           maybeAllProposals.flatTraverse { allProposals =>
-            pickMajority(allProposals).traverse { majorityHash =>
-              val newState = state.copy(status = CollectingSignatures[Artifact](majorityHash, majorityTrigger))
-              val effect = maybeProposalInfo.traverse { proposalInfo =>
-                Signature.fromHash(keyPair.getPrivate, majorityHash).flatMap { signature =>
-                  gossip.spread(ConsensusPeerDeclaration(key, MajoritySignature(signature)))
-                } >> Applicative[F].whenA(majorityHash === proposalInfo.proposalArtifactHash) {
-                  gossip.spreadCommon(ConsensusArtifact(key, proposalInfo.proposalArtifact))
-                } >> Metrics[F].recordDistribution(
-                  "dag_consensus_proposal_affinity",
-                  proposalAffinity(allProposals, proposalInfo.proposalArtifactHash)
-                )
-              }.void
-              (newState, effect).pure[F]
+            maybeLastArtifact.map { lastSignedArtifact =>
+              pickValidatedMajority(lastSignedArtifact, majorityTrigger, resources)(allProposals)
             }
+              .getOrElse(pickMajority(allProposals).pure[F])
```

Comment:
Is this the case for the first snapshot when in observing state?

Reply:
currently both cases (missing artifact or no artifact) are handled in the same way. @tbekas suggested to distinguish them but I don't know how they should be handled then

Reply:
If yes, maybe we should have to cases. One when last artifact exists and second when it doesn't. Then use `pickValidatedMajority` when `Some`, and `pickMajority` when `None`?

Reply:
Ok, so one case is when we do not store the artifact (observing state), then we can use `pickMajority`. What to do with the second case then? When the artifact should be store but we don't have it (because of async)

Reply:
fixed :+1: